state-chain/cf-integration-tests/src/fee_scaling.rs

```diff
@@ -0,0 +1,152 @@
+				buffer: 0,
+				exp_base: FixedU64::from_rational(2, 1u128),
+			});
+
+			let (mut last_gas, mut last_remaining_balance) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+
+			let mut can_pay = true;
+			(1..50).for_each(|_| {
+				match apply_extrinsic_and_calculate_gas_fee(lp, call.clone()) {
+					Ok((gas, remaining_balance)) => {
+						// Either the gas is increasing, or we've hit the ceiling of the upfront fee.
+						assert!(gas > last_gas || gas == upfront_fee);
+						assert_eq!(remaining_balance, last_remaining_balance - gas);
+						// We should never fail, and then succeed again, since the fee always
+						// increases within a block in the same pool.
+						assert!(can_pay);
+						last_remaining_balance = remaining_balance;
+						last_gas = gas;
+					},
+					Err(e) => {
+						can_pay = false;
+						// We no longer have enough to pay fees.
+						assert_eq!(
+							e,
+							TransactionValidityError::Invalid(InvalidTransaction::Payment)
+						);
+						// No balance change if transaction fails
+						assert_eq!(Flip::total_balance_of(&lp_account_id), last_remaining_balance);
+					},
+				}
+			});
+
+			// We should have run out of balance by the end.
+			assert!(!can_pay);
+		});
+}
+
+#[test]
+fn fee_scales_per_pool() {
+	const EPOCH_BLOCKS: u32 = 100;
+	const MAX_AUTHORITIES: AuthorityCount = 10;
+	let lp = AccountKeyring::Alice;
+	super::genesis::with_test_defaults()
+		.epoch_duration(EPOCH_BLOCKS)
+		.max_authorities(MAX_AUTHORITIES)
+		.with_additional_accounts(&[(
+			lp.to_account_id(),
+			AccountRole::LiquidityProvider,
+			5 * FLIPPERINOS_PER_FLIP,
+		)])
+		.build()
+		.execute_with(|| {
+			crate::network::fund_authorities_and_join_auction(MAX_AUTHORITIES);
+
+			FeeScalingRate::<Runtime>::set(FeeScalingRateConfig::ExponentBuffer {
+				buffer: 0,
+				exp_base: FixedU64::from_rational(2, 1u128),
+			});
+
+			let call = update_range_order_call(Asset::Eth);
+			let call2 = update_range_order_call(Asset::Btc);
+
+			// Different pools so fee shouldn't scale.
+			let (gas_pool1, remaining_balance_pool1) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+			let (gas_pool2, remaining_balance_pool2) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call2.clone()).unwrap();
+			assert_eq!(gas_pool1, gas_pool2);
+			assert!(remaining_balance_pool2 < remaining_balance_pool1);
+
+			// Again, once each on each pool.
+			let (gas_pool1_again, remaining_balance_pool1_again) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call.clone()).unwrap();
+			assert!(remaining_balance_pool1_again < remaining_balance_pool2);
+			let (gas_pool2_again, remaining_balance_pool2_again) =
+				apply_extrinsic_and_calculate_gas_fee(lp, call2.clone()).unwrap();
+			assert!(gas_pool1_again > gas_pool1);
+			assert!(gas_pool2_again > gas_pool2);
+			assert_eq!(gas_pool1_again, gas_pool2_again);
+			assert!(remaining_balance_pool2_again < remaining_balance_pool2);
```

Comment:
maybe we should call once more and observe the 2 pools' gas fee start to scale independently after 2 calls